webchat/utils/chat.js

```diff
@@ -0,0 +1,207 @@
+'use strict';
+(function () {
+  const app = angular.module('webchat');
+
+  app.factory('Chat', ['ChatMessage', (ChatMessage) => {
+    const Chat = class Chat {
+      /**
+       * Creates a new Chat object, which will handle a webrtc peer connection,
+       * its events and video/audio/data channels functionality.
+       * @param {MediaStream} stream - video/audio stream object
+       * @param {string} selfId - firebase id of the current user
+       */
+      constructor(stream, selfId) {
+        this.selfId = selfId;
+        this._selfStream = stream;
+        this.rpc = new RTCPeerConnection([{
+          url: 'stun:stun.l.google.com:19302',
+          url: 'stun:stun2.l.google.com:19302',
+          url: 'stun:stun3.l.google.com:19302',
+        }]);
+        this.eventHandlers = {};
+        stream.getTracks().forEach(t => {
+          this.rpc.addTrack(t, stream)
+        });
+        this.rpc.onicecandidate = e => this.emit('ice-candidate-discovered', e);
+        this.sendChannel = this.rpc.createDataChannel('sendChannel');
+        this.rpc.ondatachannel = this.handleDataChannel.bind(this);
+        this.rpc.ontrack = e => this.emit('track-received', e);
+        this.rpc.oniceconnectionstatechange = this.handleIceConnectionState.bind(this);
+        this.rpc.onsignalingstatechange = this.handleState.bind(this);
+        this._currentMessages = [];
+      }
+
+      get selfStream() {
+        return this._selfStream;
+      }
+
+      /**
+       * Mimics node's on/emit event handlers.
+       * Raises an event.
+       * @param {string} eventName - name of the event
+       * @param {object} event - event data
+       */
+      emit(eventName, event) {
+        if (this.eventHandlers[eventName]) {
+          this.eventHandlers[eventName].forEach((callback) => {
+            callback(event);
+          });
+        }
+      }
+
+      /**
+       * Mimics node's on/emit event handlers.
+       * Adds a callback to an event.
+       * @param {string} eventName - name of the event
+       * @param {Function} callback - function to be called when event is raised.
+       */
+      on(eventName, callback) {
```

Comment:
is it being used?

Reply:
These are used externally to monitor events happening and reacting to them through a callback.
i.e.:
```
            homeCtrl.client.on('call-requested', callRequestCallback);
            homeCtrl.client.on('user-list-updated', updateUserListCallback);
            homeCtrl.client.on('chat-created',  chatCreatedCallBack);
```